cli-plugins/socket/socket.go

```diff
@@ -7,24 +7,106 @@ import (
 	"io"
 	"net"
 	"os"
+	"runtime"
+	"sync"
 )
 
 // EnvKey represents the well-known environment variable used to pass the plugin being
 // executed the socket name it should listen on to coordinate with the host CLI.
 const EnvKey = "DOCKER_CLI_PLUGIN_SOCKET"
 
-// SetupConn sets up a Unix socket listener, establishes a goroutine to handle connections
-// and update the conn pointer, and returns the listener for the socket (which the caller
-// is responsible for closing when it's no longer needed).
-func SetupConn(conn **net.UnixConn) (*net.UnixListener, error) {
-	listener, err := listen("docker_cli_" + randomID())
+// NewPluginServer creates a plugin server that listens on a new Unix domain socket.
+// `h` is called for each new connection to the socket in a goroutine.
+func NewPluginServer(h func(net.Conn)) (*PluginServer, error) {
+	l, err := listen("docker_cli_" + randomID())
 	if err != nil {
 		return nil, err
 	}
 
-	accept(listener, conn)
+	if h == nil {
+		h = func(net.Conn) {}
+	}
+
+	pl := &PluginServer{
+		l: l,
+		h: h,
+	}
+
+	unlinkOnce := sync.OnceFunc(func() {
+		unlink(l)
+	})
+
+	go func() {
+		defer pl.Close()
+		for {
+			err := pl.accept()
+			unlinkOnce()
```

Comment:
This is a standard Unix behavior, which is why it's a bit under-explained. A Unix socket isn't even really a 'file' -- it's a name in the filesystem. Once the plugin has used that named path to open the other side, there is no further need to keep it around (and indeed we're trying to avoid 'leaking' socket handles in the filesystem).

A `defer` is not equivalent here as we want to unlink as early as possible. If we are killed (say, by a SIGSEGV or the OOM killer), we may end up leaking the socket.

Reply:
> ...Once the plugin has used that named path to open the other side, there is no further need to keep it around (and indeed we're trying to avoid 'leaking' socket handles in the filesystem).
> 
> A `defer` is not equivalent here as we want to unlink as early as possible. If we are killed (say, by a SIGSEGV or the OOM killer), we may end up leaking the socket.

Thanks for the explanation, makes sense! I was unaware of that unix socket behavior and the practice of removing it asap (which seemed counter intuitive to me)

Reply:
Yeah, it only makes sense in this case as we expect a single process to connect. Typically you'd unlink near program shutdown if you had a server that expected to service multiple clients (e.g. the defer).